### PR TITLE
Ability to customize SFLoginViewController

### PR DESCRIPTION
**Pull request by trooper2013**
Ability to customize SFLoginViewController. 

Overview 
1. Moved a bunch of UI related properties and behavior to overridable state in the SFLoginViewController.

2. Added a factory block setting in  SFLoginViewControllerConfig  to allow for creation of SFLoginViewController.

3. Provided the instance of LoginViewController in the AuthViewHandler block.

4. Added a subclass as example to SmartSyncExplorer.

Pending: Need to wire the authenticateAtLaunch functionality.

### CODE REVIEW

**Comment by bhariharan** [comment]
@trooper2013 Could you merge the latest from `dev`? Looks like there are some conflicts.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h [hunk 1]
```diff
@@ -98,4 +101,30 @@
  @param animated Indicates whether or not the hiding should be animated.
  */
 - (void)hideHostListView:(BOOL)animated;
+
+/** Factory Method to create the back button.
+ */
+- (UIBarButtonItem *_Nonnull)createBackButton;
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h]: Why do the create*Button methods need to be public?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h [hunk 2]
```diff
@@ -86,11 +86,8 @@
  a config object */
 @property (nonatomic, strong, nonnull) SFSDKLoginViewControllerConfig *config;
 
-/** This property will give access to the Navigation Bar */
-@property (nonatomic, strong, readonly, nullable) UINavigationBar *navBar;
-
-/** This property will give access to the SFSDKLoginHostListViewController */
-@property (nonatomic, strong, readonly, nonnull) SFSDKLoginHostListViewController *loginHostListViewController;
+/** Get the instance of nav bar. Use thi property to get the instance of navBar*/
```
Comment (bhariharan) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h]: Typo: `use this property`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h (hunk 1 above)
Comment (trooper2013) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h]: @wmathurin The create... methods should allow developers to customize the display of  3 items back button, setup button and the the Title. for e.g. the title could be a clickable item other than just a simple text.

File: native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/SFLoginExtendedViewController.m [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+- (UINavigationItem *)createTitleItem {
+    NSString *title = [SFSDKResourceUtils localizedString:@"SmartSyncLogin"];
+    // Setup top item.
+    UINavigationItem *item = [[UINavigationItem alloc] initWithTitle:title];
+    return item;
+}
+
+- (void)viewDidLoad {
```
Comment (bhariharan) [inline native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/SFLoginExtendedViewController.m]: Do we need these explicit `super` calls? Can we remove them?

File: native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/SFLoginExtendedViewController.m [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+ WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+ WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#import "SFLoginExtendedViewController.h"
+
+@interface SFLoginExtendedViewController ()
+
+@end
+
+@implementation SFLoginExtendedViewController
```
Comment (bhariharan) [inline native/SampleApps/SmartSyncExplorer/SmartSyncExplorer/Classes/SFLoginExtendedViewController.m]: I'm wondering if we even need this class. It doesn't really demonstrate much. The API is self-explanatory. All we need is the following code commented out in all our app templates:
```
SFSDKLoginViewControllerConfig *loginConfig = [[SFSDKLoginViewControllerConfig alloc] init];
    loginConfig.loginViewControllerCreationBlock = ^SFLoginViewController * _Nonnull {
        SFLoginExtendedViewController *controller = [[SFLoginExtendedViewController alloc] init];
        return controller;
    };
    [SFUserAccountManager sharedInstance].loginViewControllerConfig = loginConfig;
```

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h [hunk 5]
```diff
@@ -86,11 +86,8 @@
...
+@property (nonatomic, strong, readonly,nullable) UINavigationBar *navBar;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h]: Nit: need a space between comma and `nullable`.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h [hunk 6]
```diff
@@ -106,14 +103,28 @@
  */
 - (void)hideHostListView:(BOOL)animated;
 
-/** Determine whether or not the back button should be shown on navigation bar
+/** Factory Method to create the back button.
  */
-- (BOOL)shouldShowBackButton;
+- (UIBarButtonItem *_Nonnull)createBackButton;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController.h]: Nit: Can we use the `nonnull` preceding nomenclature here, as well as for the similar items below?  E.g.:
```obj-c
- (nonnull UIBarButtonItem *)createBackButton;
```

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFSDKLoginViewControllerConfig.h [hunk 7]
```diff
@@ -54,5 +60,6 @@
 /** Specifiy a delegate for LoginViewController. */
 @property (nonatomic, weak, nullable) id<SFLoginViewControllerDelegate> delegate;
 
+@property (nonatomic,copy) SFLoginViewControllerCreationBlock _Nullable loginViewControllerCreationBlock;
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFSDKLoginViewControllerConfig.h]: Can we put the nullability in the property attributes?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m [hunk 8]
```diff
@@ -586,6 +583,20 @@ - (void)restartAuthentication {
     [self refreshCredentials];
 }
 
+- (SFLoginViewController *) createLoginViewControllerInstance {
+    SFLoginViewController *controller = nil;
+   
+    if (self.config.loginViewControllerConfig.loginViewControllerCreationBlock) {
+        controller = self.config.loginViewControllerConfig.loginViewControllerCreationBlock();
```
Comment (khawkins) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m]: Do you feel like we may be getting a little too abstract here?  We already have the `authViewDisplayBlock` and dismiss block functionality, for a much more general case.  Should we just make the property a class of `SFLoginViewController` to instantiate?

Comment (khawkins) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFSDKOAuthClient.m]: Though it is still typed back to `SFLoginViewController`, so I guess I'm fine with either approach.

**Pull request merged**